torch/utils/pytree.py

```diff
@@ -754,14 +770,36 @@ def treespec_loads(serialized: bytes) -> PyTreeSpec:
     return treespec
 
 
-class PyTreeLeafSpecMeta(type(optree.PyTreeSpec)):  # type: ignore[misc]
-    def __instancecheck__(self, instance: object) -> bool:
-        return isinstance(instance, optree.PyTreeSpec) and instance.is_leaf()
-
-
-class PyTreeLeafSpec(optree.PyTreeSpec, metaclass=PyTreeLeafSpecMeta):
-    def __new__(cls, none_is_leaf: bool = True) -> "PyTreeLeafSpec":
-        return optree.treespec_leaf(none_is_leaf=none_is_leaf)  # type: ignore[return-value]
-
-
-LeafSpec = PyTreeLeafSpec
+if optree is not None:
+
+    class PyTreeLeafSpecMeta(type(PyTreeSpec)):  # type: ignore[misc]
+        def __instancecheck__(self, instance: object) -> bool:
+            return isinstance(instance, PyTreeSpec) and instance.is_leaf()
+
+    class PyTreeLeafSpec(PyTreeSpec, metaclass=PyTreeLeafSpecMeta):
+        def __new__(cls, none_is_leaf: bool = True) -> "PyTreeLeafSpec":
+            return optree.treespec_leaf(none_is_leaf=none_is_leaf)  # type: ignore[return-value]
+
+    LeafSpec = PyTreeLeafSpec
+
+else:
+    from ._pytree import (
+        _broadcast_to_and_flatten,
+        _register_pytree_node as register_pytree_node,
+        LeafSpec,
+        tree_all,
+        tree_all_only,
+        tree_any,
+        tree_any_only,
+        tree_flatten,
+        tree_leaves,
+        tree_map,
+        tree_map_,
+        tree_map_only,
+        tree_map_only_,
+        tree_structure,
+        tree_unflatten,
+        TreeSpec,
+        treespec_dumps,
+        treespec_loads,
+    )
```

Comment:
Note the definition of some members in `torch.utils.pytree` and `torch.utils._pytree` could be different. Users may serialize objects into files without `optree` installed and then load those files after installing `optree`. This will cause problems.

- `TreeSpec`:
	- in `torch.utils._pytree`: a dataclass
	- in `torch.utils.pytree`: an alias of `optree.PyTreeSpec`, an opaque object type. And it cannot be instantiated (users should get instances from `tree_flatten` or `tree_structure` functions).
- `treespec_dumps` and `treespec_loads`:
	- in `torch.utils._pytree`: the serialization format is a JSON string (`str`). (use `json`)
	- `torch.utils.pytree`: the serialization format is `bytes`. (use `pickle`)
	- See also: #109460